x/msgfee/decorator.go

```diff
@@ -0,0 +1,78 @@
+package msgfee
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+)
+
+// FeeDecorator implements a decorator that for each processed transaction
+// attach an additional fee to the result. Each fee is declarated per
+// transaction type. If fee is not set (zero value) then this decorator does
+// not increase the required fee value.
+// Additional fee is attached to only those transaction results that represent
+// a success.
+type FeeDecorator struct {
+	bucket *MsgFeeBucket
+}
+
+var _ weave.Decorator = (*FeeDecorator)(nil)
+
+// NewFeeDecorator returns a decorator that is upading the cost of processing
+// each message according to the fee configured per each message type.
+func NewFeeDecorator() *FeeDecorator {
+	return &FeeDecorator{
+		bucket: NewMsgFeeBucket(),
+	}
+}
+
+func (d *FeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (weave.CheckResult, error) {
+	res, err := next.Check(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	switch fee, err := txFee(d.bucket, store, tx); {
+	case err != nil:
+		return res, err
+	case !coin.IsEmpty(fee):
+		total, err := res.RequiredFee.Add(*fee)
+		if err != nil {
+			return res, errors.Wrap(err, "cannot apply message type fee")
+		}
+		res.RequiredFee = total
+	}
+	return res, nil
+}
+
+func (d *FeeDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Deliverer) (weave.DeliverResult, error) {
+	res, err := next.Deliver(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	switch fee, err := txFee(d.bucket, store, tx); {
+	case err != nil:
+		return res, err
+	case !coin.IsEmpty(fee):
+		total, err := res.RequiredFee.Add(*fee)
+		if err != nil {
+			return res, errors.Wrap(err, "cannot apply message type fee to %v")
```

Comment:
You use `%v` but there is not fmt.Sprintf, not an argument.
Maybe you want a `Wrapf()` function?
In any case, add an argument, or remove `%v`